### ISSUE TITLE
Memory leak on large datasets

### ISSUE THREAD

**Issue by SkeLLLa** [puleos/object-hash#34]
Hello.

I've experienced an issue running this lib to calculate hashes inside a loop. The hash object is quiet small ~ 6 fields, containing short arrays (from 1-6 elements):
```
{
   key: {
     subkey1: ["A", "B"],
     ...
     subkey5: ["*"]
   },
   type: 'sometype'
}
```

There are ~ 100 000 items in such array (they are coming from mongodb). If all of those keys are processed synchronously (e.g. within forEach loop) 
```
collectionData.forEach((data) => {
   ...
   data.hash = objectHash({key: data.key, type: data.type})
   ...
});
```
Object hash library eats ~ 100Mb heap size on each 5000 items and finally at ~60k it crashes node with error that says that there is not enough heap space left for JS.

After brief analysis memory snapshot shows that most of memory is consumed by `TickObject`. I think these items are produced by event emitters that's are used inside this library.

I've write a code that's enough to fit all my needs (using ES6 features):
```
const crypto = require('crypto');

/**
 * Sorts object fields
 * @param {Object} obj Initial object
 * @returns {*}
 */
const sortedObjectString = (obj) => {
  if (typeof obj == 'object') {
    if (Array.isArray(obj)) {
      return `[${obj.sort().toString()}]`;
    }
    const sortedObj = new Map();
    const keys = Object.keys(obj).sort();
    keys.forEach((key) => {
      sortedObj.set(key, sortedObjectString(obj[key]));
    });
    return `{${[...sortedObj].toString()}}`;
  }
  return obj;
};

/**
 * Calculates object hash
 * @param {Object} obj Object to hash
 * @param {string} alg Crypto algorithm to use (default="sha256");
 * @param {string=} enc Hash encoding (default="hex")
 * @returns {string} Hash string
 */
const hash = (obj, alg = 'sha256', enc = 'hex') => {
  const sorted = sortedObjectString(obj);
  return crypto.createHash(alg)
    .update(sorted)
    .digest(enc);
};
```
I know that code above doesn't cover all features of object-hash lib, but using async collection iterator for mongodb I've managed to process all collection with object-hash, but it took 36.5s to complete and eat 563Mb of memory. My code does the same thing in 6s and using 206Mb. Without using async iterator of collection object-async fails and this code processes all collection in 2.5s using 250Mb of RAM.

**Comment by addaleax** [comment]
Hi, this kind of went unnoticed here for a while. I’m having a look, I think I can reproduce this behaviour. You definitely can feel free to bump an issue here if it doesn’t get a quick response!

**Comment by addaleax** [comment]
fwiw, a fixed version is published as `object-hash@1.1.4`, even though I see that you have written other code yourself.

**Comment by SkeLLLa** [comment]
@addaleax thanks.

Yeah. I've wrote a very simple lib that supports just a small subset of various object types, but which is enough to support all objects that we're writing into db.
